// @jsr/cliffy__prompt/_generic_suggestions.js
// deno-lint-ignore-file no-explicit-any
import { bold, brightBlue, dim, stripAnsiCode, underline } from "@jsr/std__fmt/colors";
import { dirname, join, normalize } from "@jsr/std__path";
import { levenshteinDistance } from "@jsr/std__text/levenshtein-distance";
import { Figures, getFiguresByKeys } from "./_figures.js";
import { GenericInput } from "./_generic_input.js";
import { getOs } from "@jsr/cliffy__internal/runtime/get-os";
import { isDirectory } from "@jsr/cliffy__internal/runtime/is-directory";
import { readDir } from "@jsr/cliffy__internal/runtime/read-dir";
import { stat } from "@jsr/cliffy__internal/runtime/stat";
const sep = getOs() === "windows" ? "\\" : "/";
/** Generic input prompt representation. */ export class GenericSuggestions extends GenericInput {
  suggestionsIndex = -1;
  suggestionsOffset = 0;
  suggestions = [];
  #envPermissions = {};
  #hasReadPermissions;
  getDefaultSettings(options) {
    const settings = super.getDefaultSettings(options);
    return {
      ...settings,
      listPointer: options.listPointer ?? brightBlue(Figures.POINTER),
      maxRows: options.maxRows ?? 8,
      keys: {
        complete: [
          "tab"
        ],
        next: [
          "up"
        ],
        previous: [
          "down"
        ],
        nextPage: [
          "pageup"
        ],
        previousPage: [
          "pagedown"
        ],
        ...settings.keys ?? {}
      }
    };
  }
  get localStorage() {
    // Keep support for deno < 1.10.
    if (this.settings.id && "localStorage" in window) {
      try {
        // dnt-shim-ignore
        return window.localStorage;
      } catch (_) {
      // Ignore error if --location is not set.
      }
    }
    return null;
  }
  loadSuggestions() {
    if (this.settings.id) {
      const json = this.localStorage?.getItem(this.settings.id);
      const suggestions = json ? JSON.parse(json) : [];
      if (!Array.isArray(suggestions)) {
        return [];
      }
      return suggestions;
    }
    return [];
  }
  saveSuggestions(...suggestions) {
    if (this.settings.id) {
      this.localStorage?.setItem(this.settings.id, JSON.stringify([
        ...suggestions,
        ...this.loadSuggestions()
      ].filter(uniqueSuggestions)));
    }
  }
  async render() {
    if (this.settings.files && this.#hasReadPermissions === undefined) {
      // dnt-shim-ignore
      const status = await globalThis.Deno?.permissions.request({
        name: "read"
      });
      // disable path completion if read permissions are denied.
      this.#hasReadPermissions = !status || status.state === "granted";
    }
    if (this.#isFileModeEnabled()) {
      await this.#expandInputValue(this.inputValue);
    }
    await this.match();
    return super.render();
  }
  async match() {
    this.suggestions = await this.getSuggestions();
    this.suggestionsIndex = Math.max(this.getCurrentInputValue().trim().length === 0 ? -1 : 0, Math.min(this.suggestions.length - 1, this.suggestionsIndex));
    this.suggestionsOffset = Math.max(0, Math.min(this.suggestions.length - this.getListHeight(), this.suggestionsOffset));
  }
  input() {
    return super.input() + dim(this.getSuggestion());
  }
  getSuggestion() {
    return this.suggestions[this.suggestionsIndex]?.toString().substr(this.getCurrentInputValue().length) ?? "";
  }
  async getUserSuggestions(input) {
    return typeof this.settings.suggestions === "function" ? await this.settings.suggestions(input) : this.settings.suggestions ?? [];
  }
  #isFileModeEnabled() {
    return !!this.settings.files && this.#hasReadPermissions === true;
  }
  async getFileSuggestions(input) {
    if (!this.#isFileModeEnabled()) {
      return [];
    }
    const path = await stat(input).then((file)=>file.isDirectory ? input : dirname(input)).catch(()=>dirname(input));
    try {
      return await listDir(path, this.settings.files);
    } catch (error) {
      if (error instanceof Deno.errors.NotFound || error instanceof Deno.errors.PermissionDenied) {
        this.setErrorMessage(error.message);
        return [];
      }
      throw error;
    }
  }
  async getSuggestions() {
    const input = this.getCurrentInputValue();
    const suggestions = [
      ...this.loadSuggestions(),
      ...await this.getUserSuggestions(input),
      ...await this.getFileSuggestions(input)
    ].filter(uniqueSuggestions);
    if (!input.length) {
      return suggestions;
    }
    return suggestions.filter((value)=>stripAnsiCode(value.toString()).toLowerCase().startsWith(input.toLowerCase())).sort((a, b)=>levenshteinDistance((a || a).toString(), input) - levenshteinDistance((b || b).toString(), input));
  }
  body() {
    return this.getList() + this.getInfo();
  }
  getInfo() {
    if (!this.settings.info) {
      return "";
    }
    const selected = this.suggestionsIndex + 1;
    const matched = this.suggestions.length;
    const actions = [];
    if (this.suggestions.length) {
      if (this.settings.list) {
        actions.push([
          "Next",
          getFiguresByKeys(this.settings.keys?.next ?? [])
        ], [
          "Previous",
          getFiguresByKeys(this.settings.keys?.previous ?? [])
        ], [
          "Next Page",
          getFiguresByKeys(this.settings.keys?.nextPage ?? [])
        ], [
          "Previous Page",
          getFiguresByKeys(this.settings.keys?.previousPage ?? [])
        ]);
      } else {
        actions.push([
          "Next",
          getFiguresByKeys(this.settings.keys?.next ?? [])
        ], [
          "Previous",
          getFiguresByKeys(this.settings.keys?.previous ?? [])
        ]);
      }
      actions.push([
        "Complete",
        getFiguresByKeys(this.settings.keys?.complete ?? [])
      ]);
    }
    actions.push([
      "Submit",
      getFiguresByKeys(this.settings.keys?.submit ?? [])
    ]);
    let info = this.settings.indent;
    if (this.suggestions.length) {
      info += brightBlue(Figures.INFO) + bold(` ${selected}/${matched} `);
    }
    info += actions.map((cur)=>`${cur[0]}: ${bold(cur[1].join(" "))}`).join(", ");
    return info;
  }
  getList() {
    if (!this.suggestions.length || !this.settings.list) {
      return "";
    }
    const list = [];
    const height = this.getListHeight();
    for(let i = this.suggestionsOffset; i < this.suggestionsOffset + height; i++){
      list.push(this.getListItem(this.suggestions[i], this.suggestionsIndex === i));
    }
    if (list.length && this.settings.info) {
      list.push("");
    }
    return list.join("\n");
  }
  /**
   * Render option.
   * @param value        Option.
   * @param isSelected  Set to true if option is selected.
   */ getListItem(value, isSelected) {
    let line = this.settings.indent ?? "";
    line += isSelected ? `${this.settings.listPointer} ` : "  ";
    if (isSelected) {
      line += underline(this.highlight(value));
    } else {
      line += this.highlight(value);
    }
    return line;
  }
  /** Get suggestions row height. */ getListHeight(suggestions = this.suggestions) {
    return Math.min(suggestions.length, this.settings.maxRows || suggestions.length);
  }
  /**
   * Handle user input event.
   * @param event Key event.
   */ async handleEvent(event) {
    switch(true){
      case this.isKey(this.settings.keys, "next", event):
        if (this.settings.list) {
          this.selectPreviousSuggestion();
        } else {
          this.selectNextSuggestion();
        }
        break;
      case this.isKey(this.settings.keys, "previous", event):
        if (this.settings.list) {
          this.selectNextSuggestion();
        } else {
          this.selectPreviousSuggestion();
        }
        break;
      case this.isKey(this.settings.keys, "nextPage", event):
        if (this.settings.list) {
          this.selectPreviousSuggestionsPage();
        } else {
          this.selectNextSuggestionsPage();
        }
        break;
      case this.isKey(this.settings.keys, "previousPage", event):
        if (this.settings.list) {
          this.selectNextSuggestionsPage();
        } else {
          this.selectPreviousSuggestionsPage();
        }
        break;
      case this.isKey(this.settings.keys, "complete", event):
        await this.#completeValue();
        break;
      case this.isKey(this.settings.keys, "moveCursorRight", event):
        if (this.inputIndex < this.inputValue.length) {
          this.moveCursorRight();
        } else {
          await this.#completeValue();
        }
        break;
      default:
        await super.handleEvent(event);
    }
  }
  /** Delete char right. */ deleteCharRight() {
    if (this.inputIndex < this.inputValue.length) {
      super.deleteCharRight();
      if (!this.getCurrentInputValue().length) {
        this.suggestionsIndex = -1;
        this.suggestionsOffset = 0;
      }
    }
  }
  async #completeValue() {
    const inputValue = await this.complete();
    this.setInputValue(inputValue);
  }
  setInputValue(inputValue) {
    this.inputValue = inputValue;
    this.inputIndex = this.inputValue.length;
    this.suggestionsIndex = 0;
    this.suggestionsOffset = 0;
  }
  async complete() {
    let input = this.getCurrentInputValue();
    const suggestion = this.suggestions[this.suggestionsIndex]?.toString();
    if (this.settings.complete) {
      input = await this.settings.complete(input, suggestion);
    } else if (this.#isFileModeEnabled() && input.at(-1) !== sep && await isDirectory(input) && (this.getCurrentInputValue().at(-1) !== "." || this.getCurrentInputValue().endsWith(".."))) {
      input += sep;
    } else if (suggestion) {
      input = suggestion;
    }
    return this.#isFileModeEnabled() ? normalize(input) : input;
  }
  /** Select previous suggestion. */ selectPreviousSuggestion() {
    if (this.suggestions.length) {
      if (this.suggestionsIndex > -1) {
        this.suggestionsIndex--;
        if (this.suggestionsIndex < this.suggestionsOffset) {
          this.suggestionsOffset--;
        }
      }
    }
  }
  /** Select next suggestion. */ selectNextSuggestion() {
    if (this.suggestions.length) {
      if (this.suggestionsIndex < this.suggestions.length - 1) {
        this.suggestionsIndex++;
        if (this.suggestionsIndex >= this.suggestionsOffset + this.getListHeight()) {
          this.suggestionsOffset++;
        }
      }
    }
  }
  /** Select previous suggestions page. */ selectPreviousSuggestionsPage() {
    if (this.suggestions.length) {
      const height = this.getListHeight();
      if (this.suggestionsOffset >= height) {
        this.suggestionsIndex -= height;
        this.suggestionsOffset -= height;
      } else if (this.suggestionsOffset > 0) {
        this.suggestionsIndex -= this.suggestionsOffset;
        this.suggestionsOffset = 0;
      }
    }
  }
  /** Select next suggestions page. */ selectNextSuggestionsPage() {
    if (this.suggestions.length) {
      const height = this.getListHeight();
      if (this.suggestionsOffset + height + height < this.suggestions.length) {
        this.suggestionsIndex += height;
        this.suggestionsOffset += height;
      } else if (this.suggestionsOffset + height < this.suggestions.length) {
        const offset = this.suggestions.length - height;
        this.suggestionsIndex += offset - this.suggestionsOffset;
        this.suggestionsOffset = offset;
      }
    }
  }
  async #expandInputValue(path) {
    if (!path.startsWith("~")) {
      return;
    }
    const envVar = getHomeDirEnvVar();
    const hasEnvPermissions = await this.#hasEnvPermissions(envVar);
    if (!hasEnvPermissions) {
      return;
    }
    const homeDir = getHomeDir();
    if (homeDir) {
      path = path.replace("~", homeDir);
      this.setInputValue(path);
    }
  }
  async #hasEnvPermissions(variable) {
    if (this.#envPermissions[variable]) {
      return this.#envPermissions[variable];
    }
    const desc = {
      name: "env",
      variable
    };
    const currentStatus = await Deno.permissions.query(desc);
    this.#envPermissions[variable] = currentStatus.state === "granted";
    if (!this.#envPermissions[variable]) {
      this.clear();
      const newStatus = await Deno.permissions.request(desc);
      this.#envPermissions[variable] = newStatus.state === "granted";
    }
    return this.#envPermissions[variable];
  }
}
function uniqueSuggestions(value, index, self) {
  return typeof value !== "undefined" && value !== "" && self.indexOf(value) === index;
}
async function listDir(path, mode) {
  const fileNames = [];
  for (const file of (await readDir(path))){
    if (mode === true && (file.name.startsWith(".") || file.name.endsWith("~"))) {
      continue;
    }
    const filePath = join(path, file.name);
    if (mode instanceof RegExp && !mode.test(filePath)) {
      continue;
    }
    fileNames.push(filePath);
  }
  return fileNames.sort(function(a, b) {
    return a.toLowerCase().localeCompare(b.toLowerCase());
  });
}
function getHomeDirEnvVar() {
  return Deno.build.os === "windows" ? "USERPROFILE" : "HOME";
}
function getHomeDir() {
  return Deno.env.get(getHomeDirEnvVar());
}
