// @openneuro/cli/src/commands/git-credential.js
import { Command } from "@jsr/effigies__cliffy-command";
import { TextLineStream } from "@jsr/std__streams";
import { getConfig } from "../config.js";
const prepareRepoAccess = `
  mutation prepareRepoAccess($datasetId: ID!) {
    prepareRepoAccess(datasetId: $datasetId) {
      token
      endpoint
    }
  }
`;
export async function getRepoAccess(datasetId) {
  const config = getConfig();
  const req = await fetch(`${config.url}/crn/graphql`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "Authorization": `Bearer ${config.token}`
    },
    body: JSON.stringify({
      query: prepareRepoAccess,
      variables: {
        datasetId
      }
    })
  });
  const response = await req.json();
  if (response.errors) {
    const errors = response.errors;
    throw Error(errors.map((error)=>error.message).toString());
  } else {
    return {
      token: response.data.prepareRepoAccess.token,
      endpoint: response.data.prepareRepoAccess.endpoint
    };
  }
}
/**
 * Provide a git-credential helper for OpenNeuro
 */ export async function gitCredentialAction(stdinReadable = Deno.stdin.readable, tokenGetter = getRepoAccess) {
  let pipeOutput = "";
  const credential = {};
  // Create a stream of lines from stdin
  const lineStream = stdinReadable.pipeThrough(new TextDecoderStream()).pipeThrough(new TextLineStream());
  for await (const line of lineStream){
    const [key, value] = line.split("=", 2);
    credential[key] = value;
  }
  if ("path" in credential && credential.path) {
    const datasetId = credential.path.split("/").pop();
    const { token } = await tokenGetter(datasetId);
    const output = {
      username: "@openneuro/cli",
      password: token
    };
    for(const key in output){
      pipeOutput += `${key}=${output[key]}\n`;
    }
  } else {
    throw new Error("Invalid input from git, check the credential helper is configured correctly");
  }
  return pipeOutput;
}
export const gitCredential = new Command().name("git-credential").description("A git credentials helper for easier datalad or git-annex access to datasets.")// Credentials here are short lived so store is not useful
.command("store").action(()=>{}).command("get").action(async ()=>{
  console.log(await gitCredentialAction());
});
