// @openneuro/cli/src/options.ts
import { Command, EnumType } from "@jsr/effigies__cliffy-command"
import { type LevelName, LogLevelNames } from "@jsr/std__log"
import type { ValidatorOptions } from "jsr:/@bids/validator@^2.2.1/options"
import denoJson from "../deno.json" with { type: "json" }
import { setupLogging } from "./logger.js"
import { login } from "./commands/login.js"
import { upload } from "./commands/upload.js"
import { download } from "./commands/download.js"
import { gitCredential } from "./commands/git-credential.js"
import { specialRemote } from "./commands/special-remote.js"
import { createDatasetCommand } from "./commands/create-dataset.js"

export type OpenNeuroOptions = {
  datasetPath: string
  localPath?: string
  validatorOptions?: ValidatorOptions
  debug?: LevelName
}

const openneuroCommand = new Command()
  .name("openneuro")
  .description(
    "OpenNeuro command line tools for uploading, downloading, or syncing datasets. See https://docs.openneuro.org for detailed guides.",
  )
  .version(denoJson.version)
  .globalType("debugLevel", new EnumType(LogLevelNames))
  .globalEnv("OPENNEURO_LOG=<type:debugLevel>", "Enable debug output.")
  .globalAction(({ openneuroLog }) => {
    setupLogging(openneuroLog ? openneuroLog : "ERROR")
  })
  .globalEnv("OPENNEURO_API_KEY=<key:string>", "Specify an OpenNeuro API key.")
  .command("login", login)
  .command("download", download)
  .command("upload", upload)
  .command("git-credential", gitCredential)
  .command("special-remote", specialRemote)
  .command("create", createDatasetCommand)

/**
 * Parse command line options and return a OpenNeuroOptions config
 * @param argumentOverride Override the arguments instead of using Deno.args
 */
export async function commandLine(
  argumentOverride: string[],
): Promise<OpenNeuroOptions> {
  const { args, options } = await openneuroCommand.parse(argumentOverride)

  return {
    datasetPath: args[0] as string,
    ...options,
  }
}
